jboss/container/wildfly/launch/os/node-name/added/launch/openshift-node-name.sh

```diff
@@ -8,9 +8,12 @@ function init_node_name() {
       JBOSS_NODE_NAME="${HOSTNAME}"
     fi
 
-    # CLOUD-427: truncate to 23 characters max (from the end backwards)
+    # CLOUD-2268: truncate to 23 characters max
     if [ ${#JBOSS_NODE_NAME} -gt 23 ]; then
-      JBOSS_NODE_NAME=${JBOSS_NODE_NAME: -23}
+      local prefix=${JBOSS_NODE_NAME::14}
+      local suffix=${JBOSS_NODE_NAME#$prefix}
+      suffix=$(echo $suffix | md5sum | cut -c -8 )
```

Comment:
This seems OK to me, you would like need a lot of pods running before you'd get a collision.

Reply:
@ochaloup , I am implementing something similar for the bootable JAR, until now I was using the last 23 characters. Seemed to me large enough. In which use-case we need your new fix? It seems that I would need to have that too.

Reply:
@jfdenise yes, until now this script was working on stripping of the node name to last 23 characters.
You are right that taking the last 23 characters could be unique enough, or it could not be. It depends on naming patterns used by every user.
This fix tries to increase the probability that the clash of the names won't happen (because when the name clash after stripping happens the two pods running in parallel and using the same database will start causing transaction inconsistency).
This fix is based on discussion in Narayana team and especially on points from @tomjenkinson where stripping was not considered as good enough. As I have no data about naming patterns that users use for pod naming it's hard to say for me "it's"/"it is not" large enough.

about the bootable jar: yes. If we (me, you, @luck3y , @tomjenkinson...) agree that this fix makes the probability of clash in naming lower and thus it makes sense the similar handling will be needed for bootable JAR if it sets the transaction node id somehow. I assume it would be good to have a similar mechanism in all WFLY "deployments".

Reply:
@ochaloup , thank-you. So we would be in a case where the POD name is not generated by Openshift (generate-name: myapp-) but explicitly set by the user (name=myapp-pod1). When generate-name is used, the logic present in master branch seems good enough, if a user controls the POD names, we  need your new fix.

Reply:
@ochaloup If you're focused solely on reducing collisions, dropping the '-' gives you one more byte. It makes it harder to understand what is going on though, but it is "safer." I think either is fine, as I mentioned you'd need a *lot* of pods before truncated md5 collisions became an issue.

Reply:
@jfdenise I see. I understand your rationale. I would only add that the OpenShift never (as far as I'm aware) generates the pod name on its own. It's created from the application name. In case the company has some specific specific naming pattern (e.g. adds some suffix to each application) then there could be some clash more easily.

Anyway, what is the final outcome for this PR? Do you think this adjustment makes sense and could be merged or it's not appropriate by your opinion?

Reply:
@ochaloup , I was experimenting with your fix and realize that a newline is added by echo prior to be passed to md5sum, you could perhaps use echo -n

Reply:
@ochaloup to answer on the fix in itself, my understanding is that MD5 generates uniform random numbers, so truncating to 8 hexa characters should give us the chance of a collision every 16^8 occurrences. Do you agree? This seems large enough.

Reply:
@jfdenise oh, that's right. thanks for the point with the `echo -n`. I didn't realize that (updated).

Regarding the chance of the collision.
My way to calculate this will go like this: there are truncated digits of the hexadecimal number. The number represents the 128 bits (128 long binary number). What I understand the `md5` has the possibility of collision being 2^64 (2^128 possibilities but it seems that because of the birthday problem the chance for collision is 2^64). In respect of the uniform distribution of the hash function then naively I would say I stripping it to 8 hex numbers means ~32 bits to be used which would mean chance of collision being 2^16.

Reply:
Ah, ok, same reasoning but didn't think at the birthday problem...so less chance of collision 1/65560. Can't we have a longer suffix of 10 or even 12 chars?

Reply:
@jfdenise @ochaloup I suspect the real fix here is really to address the underlying 23 byte restriction itself rather than improve the workaround here. I think this change is fine as it significantly reduces the issue, but the issue will always be that truncating md5 or sha will allow collisions. (md5 is 32 bytes I think, isn't it?)

Note I have no idea if the 23 byte node name restriction is something we can do anything about either...

Reply:
For completeness here, xid is 20 bytes, so could perhaps be an alternative to this: (https://github.com/rs/xid) but I think this is still significantly better to what we have now, so I'm just leaving this here as a note.

Reply:
@jfdenise @luck3y thanks a lot for your points. I made the proposed change to use the 12 chars for the hash and used the tip of removing the `-` character.

It's a good point about the Xid format. I would add this.
The shape of the `Xid` used here is the "XA Xid" and is defined by java API https://docs.oracle.com/javase/7/docs/api/javax/transaction/xa/Xid.html#getGlobalTransactionId() and it consist of 64 bytes (https://docs.oracle.com/javase/7/docs/api/constant-values.html#javax.transaction.xa.Xid.MAXBQUALSIZE).
But the "XA `Xid`" byt array does not contain only the node identifier - which is defined to be 23 long. This is proprietary decision of Narayana and hardly could be changed. The format of the byte array is defined by Narayana whole and it's used to save other information which are necessary for txn processing.
Now regarding the unique generator of the xid. That's not so easy here as we can't use as simply as it is - at least in current Narayana. Narayana does not store the node identifier persistently and it receives the value only during startup. Based on the identifier it knows what transaction is responsible for.
As the identifier is passed externally we require the guarantee that the same node with the same object store provides the same node identifier again for Narayana. That's why we couple the node identifier with the "host name/pod name". That's the reason Narayana transaction processing requires the use of the statefulset where the pod name stability to persistence volume stability is guaranteed.
It's an interesting topic if this limitation could be adjusted. I will check with colleagues.

Reply:
@ochaloup Ok, thanks. I wouldn't want that to delay this change going in either, as this improves the situation in any case.

Reply:
@ochaloup Is this OK to merge as-is now?